### ISSUE TITLE
Creating arbitrary  hash on lonely collection and encapsulating collection items

### ISSUE THREAD

**Issue by konung** [apotonick/roar#111]
Hello. 

Trying to figure out how to send some arbitrary data on a lonely collection. I tried every which way and it's either overwriting items or just get ignored

Have something like

````ruby
# representers/product_representer.rb
class ProductRepresenter  < Roar::Decorator
  include Roar::Representer::JSON

  property :id
  property :name
  property :created_at
  property :updated_at
end

# representers/products_representer.rb
class ProductsRepresenter < Roar::Decorator
  include Representable::JSON::Collection

  items extend: ProductRepresenter, class: Product
end

````

This works find and produces basic output when I make call to /products.json

However I would like to add some arbitrary attributes to the collection

so right now I get

````json
[{
"id": 51,
"name": "product 1",
"created_at": "2008-09-15T09:37:12.000-05:00",
"updated_at": "2013-12-12T17:04:28.000-06:00",
},
{
"id": 52,
"name": "product 2",
"created_at": "2008-09-15T09:37:12.000-05:00",
"updated_at": "2013-12-12T17:28:45.000-06:00",
},
.....
]
````

I'd like to be able to spit out something like this separate items (products) and add meta attributes:

````json
{
"meta" : {
"ordered_by": "asc",
"request_made_at" : "",
"total_records" : "4000"
},
"products" : [
{
"id": 51,
"name": "product 1",
"created_at": "2008-09-15T09:37:12.000-05:00",
"updated_at": "2013-12-12T17:04:28.000-06:00",
},
{
"id": 52,
"name": "product 2",
"created_at": "2008-09-15T09:37:12.000-05:00",
"updated_at": "2013-12-12T17:28:45.000-06:00",
},
.....]
}
````

These are just some examples. The idea being that the API responses I'm writing require some extensive meta data included. Right now I'm doing that just using plain old Rails json responses and templates, but would like to switch to roar.

**Comment by ichthala** [comment]
I just wrestled with a similar issue for a while. I had a representer like this:

```
class EmailAddressCollectionRepresenter < Roar::Decorator
  include Roar::Representer::JSON::HAL
  include Representable::JSON::Collection

  items decorator: EmailAddressRepresenter, class: Channel

  link :self do |opts|
    "#{opts[:path]}/email_addresses"
  end

end
```

...and I was extremely puzzled as to why the link was being omitted when the resource was accessed. It's pretty confusing that `collection` allows extra properties at the same level as the collection's namespace, but `items` doesn't. Even if this isn't something you want to support, it might be helpful to make it clearer in the documentation. (Maybe in the documentation for Representable rather than Roar.) I'm still not 100% clear on why it's not supported...

**Comment by apotonick** [comment]
The `link` is omitted because you're using a "lonely collection", which solely consists of items in a collection. The whole point of a lonely collection is that there's no other elements in it.

@ichthala, why don't you use a normal representer? Try the following

```ruby
class EmailAddressCollectionRepresenter < Roar::Decorator
  include Roar::Representer::JSON::HAL
  include Representable::JSON::Collection

  collection: :email_addresses, decorator: EmailAddressRepresenter, class: Channel

  link :self do |opts|
    "#{opts[:path]}/email_addresses"
  end

  def email_addresses
    represented
  end
end
```

**Comment by ichthala** [comment]
Thanks for the help! That almost worked, but I had to do this:

```
class EmailAddressCollectionRepresenter < Roar::Decorator
  include Roar::Representer::JSON::HAL
  # include Representable::JSON::Collection

  collection: :email_addresses, decorator: EmailAddressRepresenter, class: Channel, exec_context: :decorator

  link :self do |opts|
    "#{opts[:path]}/email_addresses"
  end

  def email_addresses
    represented
  end
end
```

It looks like if you `include Representable::JSON::Collection`, it will omit any other properties besides the collection. Is that accurate? Is that module only supposed to be for lonely collections?

**Comment by apotonick** [comment]
Yeah, you're right! I'm an idiot, I was telling you _not_ to use `Collection` and then left it in the code snippet... :stuck_out_tongue_winking_eye:

**Comment by ichthala** [comment]
Haha, alright, just making sure. But I'm confused -- is that module only for lonely collections? If so perhaps it could be renamed?

**Comment by apotonick** [comment]
This module, yes, is only for lonely collections, and, no, I'm not gonna rename it, as it is well documented here: https://github.com/apotonick/representable/#lonely-collections :stuck_out_tongue_closed_eyes: 

Using the `Collection` module in arbitrary representers is not mentioned anyway, what made you think this is how it works, if I may ask?

**Comment by ichthala** [comment]
I didn't think that, I was including the module specifically to use `items`. Ultimately, I wanted to pass in an array of objects and have them be represented without having to wrap the array in another object just for the representer. `items` provided that pattern, and I was pleased, so I started building off of it and then went astray with the namespacing and links. It wasn't obvious (though in retrospect it should have been) that I could use `collection` and still not have to wrap the array. ...I hope that makes sense/answers your question. Thanks again for all the clarifications. ^_^

**Comment by apotonick** [comment]
Maybe we should note in the README that as soon as you want "more" (e.g. hypermedia) in a lonely collection, you should switch to a non-lonely representer?

**Comment by ichthala** [comment]
Yeah, I think that would make it easier for people to quickly figure out
which feature they need.
